### ISSUE TITLE
Unify modeling API around one class per model type

### ISSUE THREAD

**Issue by jbischof** [keras-team/keras-nlp#377]
# Problem

Our current modeling API avoids creating an explicit config system by using config-in-code classes. These call a generic constructor with hardcoded arguments. In practice this requires us to maintain a bloated set of symbols in order to load a wide variety of checkpoints. If a user wants to add their own checkpoint with a different configuration they first need to create another config-in-code class.

```python
# The powerful API that we don"t use directly
# Does not allow checkpoint loading
keras_nlp.models.BertCustom(
    vocabulary_size=30522,
    num_layers=12,
    num_heads=12,     
    hidden_dim=768,
    intermediate_dim=3072,
    max_sequence_length=12,
    name="encoder",
)
# The API used by the rest of the library is a thin wrapper for the call above
keras_nlp.models.BertBase(weights="bert_base_uncased_en")
```

While this ostensibly avoids a config system, we reference this these thin wrappers in checkpoint metadata to essentially inline the config:
```python
checkpoints = {
    "bert_base_uncased_en": {
        "model": "BertBase",          # THIS IS THE NAME OF A HARDCODED CONFIG
        "vocabulary": "uncased_en",
        "description": (
            "Base size of BERT where all input is lowercased. "
            "Trained on English Wikipedia + BooksCorpus."
        ),
        "weights_url": "https://storage.googleapis.com/keras-nlp/models/bert_base_uncased_en/model.h5",
        "weights_hash": "9b2b2139f221988759ac9cdd17050b31",
    },
}
```

The result is a loss of cohesion and flexibility in the library. `BertCustom` is clearly the powerful API but we all but hide it from users and don"t make it easy to load our checkpoints or make their own checkpoints with it. All configs must be anticipated by the package authors and the users have to navigate a large number of classes for example:
```python
# Why different classes? Why should the user care?
BertBase("bert_base_uncased_en")
BertLarge("bert_large_uncased_en")
```

This gets especially silly when we think of the `BertClassifier` model. We want to offer checkpoints for sentiment classifiers (#375), but in order to avoid providing configuration information we would have to create a config-in-code class for every possible number of classes!

```python
classifier_checkpoints = {
    "toxic_bert": {
        "model": "BertClassifierWith6Classes",          # THIS IS SILLY
        "backbone_model": "BertBase"
        "vocabulary": "uncased_en",
        "description": (
            "A multi-headed model capable of detecting different types "
            "of toxicity like threats, obscenity, insults, and identity-based hate."
        ),
        "weights_url": "https://storage.googleapis.com/keras-nlp/models/toxic_bert.h5",
        "weights_hash": "9b2b2139f221988759ac9cdd17050b31",
    },
}
# WHY!?!?!?
BertClassifierWith6Classes(weights="toxic_bert")
```

# Proposal

We can unify the API around one flexible, cohesive symbol using the native `keras.Model` methods of `from_config` and `to_config`.  With this functionality we do not need a maintain a parallel set of config classes. The config is just the constructor signature in most cases, and advanced functionality can be added by explicitly overriding `from_config`.

Now the same symbol can be called by all users:
```python
# Random graph
keras_nlp.models.Bert(
    vocabulary_size=30522,
    num_layers=12,
    num_heads=12,     
    hidden_dim=768,
    intermediate_dim=3072,
    max_sequence_length=12,
    name="encoder"
)
# Pretrained graph
keras_nlp.models.Bert.from_checkpoint("bert_base_uncased_en")
```

The checkpoint structs will now have the native Keras config for the model:
```python
#===> bert.py:
checkpoints = {
    "bert_base_uncased_en": {
        "config": {
            "vocabulary_size": 10000, 
            "hidden_dim": 768, 
            "intermediate_dim": 3072, 
            "num_layers": 12, 
            "num_heads": 12, 
            "max_sequence_length": 512, 
            "num_segments": 2, 
            "dropout": 0.1
        },
        "vocabulary": "uncased_en",
        "description": (
            "Base size of BERT where all input is lowercased. "
            "Trained on English Wikipedia + BooksCorpus."
        ),
        "weights_url": "https://storage.googleapis.com/keras-nlp/models/bert_base_uncased_en/model.h5",
        "weights_hash": "9b2b2139f221988759ac9cdd17050b31",
    },
}

#===> bert_tasks.py
classifier_checkpoints = {
    "toxic_bert": {
        "config": {
            "backbone_config": {
                "vocabulary_size": 10000, 
                "hidden_dim": 768, 
                "intermediate_dim": 3072, 
                "num_layers": 12, 
                "num_heads": 12, 
                "max_sequence_length": 512, 
                "num_segments": 2, 
                "dropout": 0.1
        },
            "num_classes": 6,
        },
        "description": (
            "A multi-headed model capable of detecting different types "
            "of toxicity like threats, obscenity, insults, and identity-based hate."
        ),
        "weights_url": "https://storage.googleapis.com/keras-nlp/models/toxic_bert.h5",
        "weights_hash": "9b2b2139f221988759ac9cdd17050b31",
    },
}
```

The main difference between `Bert` and the old `BertCustom` API is the ability to load checkpoints using a `classmethod`:

```python
class Bert(keras.Model):
    ...
    @classmethod
    def from_checkpoint(checkpoint="bert_base_uncased_en") :
        model = Bert.from_config(checkpoints[name]["config"])
        model.load_weights(...)
...
```

Here `BertClassifier` constructor will work similarly, although we will have to implement a custom `from_config` to handle the nested structure:
```python
class BertClassifier(keras.Model):
    ...
    @classmethod
    def from_checkpoint(checkpoint="toxic_bert") :
        model = BertClassifier.from_config(classifier_checkpoints[name]["config"])
        model.load_weights(...)
...
```

# Architecture aliases
Since model architectures generally conform to a small number of templates, we can maintain a struct of hardcoded configs instead of hardcoded classes. In this setup the `checkpoints` struct would look identical but we would maintain another struct of architectures:

```python
architectures = {
    "bert_base" : {
        "vocabulary_size": 10000, 
        "hidden_dim": 768, 
        "intermediate_dim": 3072, 
        "num_layers": 12, 
        "num_heads": 12, 
        "max_sequence_length": 512, 
        "num_segments": 2, 
        "dropout": 0.1
    },
}
```

The checkpoint structs are now significantly cleaner:
```python
#===> bert.py:
checkpoints = {
    "bert_base_uncased_en": {
        "config": "bert_base",
        "vocabulary": "uncased_en",
        "description": (
            "Base size of BERT where all input is lowercased. "
            "Trained on English Wikipedia + BooksCorpus."
        ),
        "weights_url": "https://storage.googleapis.com/keras-nlp/models/bert_base_uncased_en/model.h5",
        "weights_hash": "9b2b2139f221988759ac9cdd17050b31",
    },
}

#===> bert_tasks.py
classifier_checkpoints = {
    "toxic_bert": {
        "config": {
            "backbone_config": "bert_base"
            "num_classes": 6,
        },
        "description": (
            "A multi-headed model capable of detecting different types "
            "of toxicity like threats, obscenity, insults, and identity-based hate."
        ),
        "weights_url": "https://storage.googleapis.com/keras-nlp/models/toxic_bert.h5",
        "weights_hash": "9b2b2139f221988759ac9cdd17050b31",
    },
}
```

This would also allow us to set up convenience constructors for random graphs such as
```python
Bert.from_architecture(
    architecture="bert_base"
    vocabulary_size=10000
)
```

# User supplied weights
Since many power users will not want to rely on our GCP bucket to load weights, we should provide a `weights` argument that overrides our RCP weight loading but still uses our `checkpoints` config for all other information. If the supplied weights are incompatible with the constructed graph we can throw a `ValueError`.

```python
# User supplied weights for entire checkpoint---no RCPs involved!
keras_nlp.models.Bert.from_checkpoint(
    checkpoint="bert_base_uncased_en",
    weights="./my_weights.h5",
)
# User supplied weights for arch constructor.
# Interesting for consistency but less important since no RCPs to avoid
keras_nlp.models.Bert.from_architecture(    
    architecture="bert_base",
    vocabulary_size=10000,
    weights="./my_weights.h5",
)
# Equivalent to above, also for consistency
keras_nlp.models.Bert(
    weights="./my_weights.h5",
    vocabulary_size=30522,
    num_layers=12,
    num_heads=12,     
    hidden_dim=768,
    intermediate_dim=3072,
    max_sequence_length=12,
    name="encoder",
)
```

Note that in this design `weights` takes on a much more consistent meaning as "user-supplied weights" and does not allow for checkpoint or architecture overloading.

# API summary
Here are examples of the different API surfaces for interacting with a BERT model. Note that all can take a `weights` argument which pertains to a local file of compatible weights only.

```python
# Random encoder
keras_nlp.models.Bert(
    vocabulary_size=30522,
    num_layers=12,
    num_heads=12,     
    hidden_dim=768,
    intermediate_dim=3072,
    max_sequence_length=12,
)
# Random encoder from architecture alias
keras_nlp.models.Bert.from_architecture(
    architecture="bert_base",
    vocabulary_size=30522,    
)
# Pretrained encoder checkpoint
keras_nlp.models.Bert.from_checkpoint(
    checkpoint="bert_base_uncased_en",
)
# Classifier for fine-tuning
keras_nlp.models.BertClassifier(
    backbone="bert_base_uncased_en",
    num_classes=6
)
# Pretrained classifier
keras_nlp.models.BertClassifier.from_checkpoint(
    checkpoint="toxic_bert",
)
```

# Drawbacks
The biggest drawback to this proposal is that we are deviating significantly from the `keras.applications` and `keras_cv` style by using `classmethod` constructors rather than having special handling in `__init__` if `weights` is not `None`. It is technically possible to create the functionality described above using `__init__` and `__new__` but it will lead to meaningfully more complicated code. We are also deviating by avoiding the hardcoded classes, but I hope that this is a usability win.

**Comment by jbischof** [comment]
While my `BertClassifier` proposal technically works I just realized that we are reusing `bert_base_uncased` to get a `BertBase` arch, which is a bit sloppy and doesn't offer the advertised flexibility. Two alternatives are adding `bert_base_random` to `bert.checkpoints` (clean but not flexible), or somehow allowing a config specification for `backbone` or nested config overall. Ideas welcome!

**Comment by mattdangerw** [comment]
Thanks very much for the writeup!

I think we may be missing a bigger difference between this proposal and the current API, besides just consistency (though consistency is an important consideration too). In the `keras.applications` mental model, our checkpoint IDs are just "shortcuts" to specify certain parameters directly. It is very illuminating to imagine taking a flow currently from our GCS checkpoints to user-hosted data.

### Current API

**Keras hosted flow**
```python
# Preprocess data.
dataset = load_my_data()
preprocessor = keras_nlp.models.BertPreprocessor(
    vocabulary="bert_base_multi_cased",
)
dataset = dataset.map(lambda x, y: preprocessor(x), y)
# Define graph.
encoder = keras_nlp.models.BertBase(
    weights="bert_base_multi_cased",
)
classifier = keras_nlp.models.BertClassifier(model, 2)
# Fine tune.
classifier.compile(...)
classifier.fit(dataset)
```

**User hosted flow**
```python
# Preprocess data.
dataset = load_my_data()
preprocessor = keras_nlp.models.BertPreprocessor(
    vocabulary="./path/to/a/vocab.txt",
)
dataset = dataset.map(lambda x, y: preprocessor(x), y)
# Define graph.
encoder = keras_nlp.models.BertBase(
    weights="./path/to/a/weights.h5",
    vocabulary_size=preprocessor.vocabulary_size(),
)
classifier = keras_nlp.models.BertClassifier(model, 2)
# Fine tune.
classifier.compile(...)
classifier.fit(dataset)
```
 
In the `keras.applications` mental model, you are always sitting on an API where can easily bring your own data. That's basically forced by the fact that you can sub in an "ID shortcut" for a weights file or a vocab file. I think this view can also help explain some of the KerasCV APIs, like RetinaNet, to some extent.

### Proposed API

Under the new proposal, brining your own data looks dramatically different.

**Keras hosted flow**
```python
# Preprocess data.
dataset = load_my_data()
preprocessor = keras_nlp.models.BertPreprocessor.from_checkpoint("bert_base_uncased_en")
dataset = dataset.map(lambda x, y: preprocessor(x), y)
# Define graph.
encoder = keras_nlp.models.Bert.from_checkpoint("bert_base_uncased_en")
classifier = keras_nlp.models.BertClassifier(model, 2)
# Fine tune.
classifier.compile(...)
classifier.fit(dataset)
```

**User hosted flow with new API**
```python
# Preprocess data.
dataset = load_my_data()
preprocessor = keras_nlp.models.BertPreprocessor(
    vocabulary="./path/to/a/vocab.txt",
)
dataset = dataset.map(lambda x, y: preprocessor(x), y)
# Define graph.
encoder = keras_nlp.models.Bert(
    vocabulary_size=preprocessor.vocabulary_size(),
    hidden_dim=768, 
    intermediate_dim=3072, 
    num_layers=12, 
    num_heads=12, 
    max_sequence_length=512, 
    num_segments=2, 
    dropout=0.1
)
encoder.set_weights("./path/to/a/weights.h5")
classifier = keras_nlp.models.BertClassifier(model, 2)
# Fine tune.
classifier.compile(...)
classifier.fit(dataset)
```

For a project like huggingface, the hub seems like the business model. Having a separate and super easy API surface that you can tap into if you upload to their hub is great. The fact that you need to conform to their data format to use their static construction API fits well with that positioning.

For us, I worry we would be building an API that screams "hub!" without a real plan of action.

Under the proposal, a substantial piece of our API requires using not just GCS, but our specific GCS bucket. We already uploaded some weights files that are many gigs. Are we prepared to vouch for download bandwidth to some startup, that needs to run XXX jobs a day?

We could look at defining a file format so you could run `model.from_checkpoint("./path/to/my/checkpoint/")`, but then we are literally rebuilding Keras saved model (where there is a lot of work going already!). That's a rabbit hole worth discussing but probably beyond a github issue thread.

**Comment by mattdangerw** [comment]
If we did want to stick close to the existing UX, I do think there are some alternatives where we could consolidate the pertained `Bert` into a single symbol without a static construction offering. The obvious one seems like providing an `architecture` argument, that is inferred when `weights=id` is given, and otherwise not. `BertBackbone(architecture="base", vocabulary_size=10_000, weights="file.h5")` or `BertBackbone(weights="bert_base_uncased_en")`. And then `BertCustomBackbone(num_layers...)`. That might be exacerbating the problem you are trying to solve though.

For `BertClassifierWith6Classes` the obvious alternative there seems to be to keep that piece of config in our metadata, and set it automatically, when `weights=id`. `BertClassifier(weights="some-file.h5", num_classes=5)` or `BertClassifier(weights="some-id-with-5-classes")`

**Comment by jbischof** [comment]
Thanks for the important observation about user-supplied weights @mattdangerw! I've updated the proposal to include a `weights` argument that only pertains to user-supplied weights. We now see that the user journey only differs by a single parameter.
